### PR TITLE
New Guard Authentication System (e.g. putting the joy back into security)

### PR DESCRIPTION
**Pull request by weaverryan**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | at least partially: #14300, #11158, #11451, #10035, #10463, #8606, probably more |
| License | MIT |
| Doc PR | symfony/symfony-docs#5265 |

Hi guys!

Though it got much easier in 2.4 with `pre_auth`, authentication is a pain in Symfony. This introduces a new authentication provider called guard, with one goal in mind: put everything you need for _any_ authentication system into one spot. 
### How it works

With guard, you can perform custom authentication just by implementing the [GuardAuthenticatorInterface](https://github.com/weaverryan/symfony/blob/guard/src/Symfony/Component/Security/Guard/GuardAuthenticatorInterface.php) and registering it as a service. It has methods for every part of a custom authentication flow I can think of.

For a working example, see https://github.com/weaverryan/symfony-demo/tree/guard-auth. This uses 2 authenticators simultaneously, creating a system that handles [form login](https://github.com/weaverryan/symfony-demo/blob/guard-auth/src/AppBundle/Security/FormLoginAuthenticator.php) and [api token auth](https://github.com/weaverryan/symfony-demo/blob/guard-auth/src/AppBundle/Security/TokenAuthenticator.php) with a respectable amount of code. The [security.yml](https://github.com/weaverryan/symfony-demo/blob/guard-auth/app/config/security.yml) is also quite simple.

This also supports "manual login" without jumping through hoops: https://github.com/weaverryan/symfony-demo/blob/guard-auth/src/AppBundle/Controller/SecurityController.php#L45

I've also tested with "remember me" and "switch user" - no problems with either.

I hope you like it :).
### What's Needed

1) **Other Use-Cases?**: Please think about the code and try it. What use-cases are we _not_ covering? I want Guard to be simple, but cover the 99.9% use-cases.

2) **Remember me** functionality cannot be triggered via manual login. That's true now, and it's not fixed, and it's tricky.
### Deprecations?

This is a new feature, so no deprecations. But, creating a login form with a guard authenticator is a whole heck of a lot easier to understand than `form_login` or even `simple_form`. In a perfect world, we'd either deprecate those or make them use "guard" internally so that we have just **one** way of performing authentication.

Thanks!

### CODE REVIEW

File: src/Symfony/Component/Security/Guard/Provider/GuardAuthenticationProvider.php [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+            throw new \UnexpectedValueException(sprintf(
+                'The %s::authenticate method must return a UserInterface. You returned %s',
+                get_class($guardAuthenticator),
+                is_object($user) ? get_class($user) : gettype($user)
+            ));
+        }
+
+        // check the AdvancedUserInterface methods!
+        $this->userChecker->checkPreAuth($user);
+        $this->userChecker->checkPostAuth($user);
```
Comment (iltar) [inline src/Symfony/Component/Security/Guard/Provider/GuardAuthenticationProvider.php]: Shouldn't the `checkPostAuth()` be done _after_ `createAuthenticatedToken()`?

I have a small concern regarding the user checker though. Say that before querying anything to find a user, I want to check if the IP has a time out, how would I do it here?

In the old situation I would have created my own simple-authenticator which did this. After this PR, I'd have to put it in a `GuardAuthenticatorInterface`. Neither of those solutions seems feasible for me as I would have to create either a weird inheritance tree or duplicate code in my authenticators. It would be great for you (and others ofc!) to think with me on the following points:
 - Would it be possible to use an actual FormType (with optional validation) here?
 - Would it be possible to implement something related to #11090?

Regarding those points, after an x amount of failed attempts, I have to show a captcha which has to be displayed until either valid or another x attempts have failed. This bundle would add the 'captcha' type to a form which you can then re-use for both the validation and display. I want to do this _before_ the authentication takes place (why query the database for a blocked IP). https://github.com/Gregwar/CaptchaBundle

_Just a random brainfart_
I can also imagine the possibility to automate certain validation points by using a FormType with constraints; With a data transformer to put the user in there automatically via a user provider based on the username.

File: src/Symfony/Component/Security/Guard/Token/GuardTokenInterface.php [hunk 2]
```diff
@@ -0,0 +1,15 @@
+<?php
+
+namespace Symfony\Component\Security\Guard\Token;
+
+/**
+ * An empty interface that both guard tokens implement.
+ *
+ * This interface is used by the GuardAuthenticationProvider to know
+ * that a token belongs to its system.
```
Comment (iltar) [inline src/Symfony/Component/Security/Guard/Token/GuardTokenInterface.php]: An interface that knows about its implementations? ;)

What about replacing `An empty interface` by `A marker interface`?

File: src/Symfony/Component/Security/Guard/Token/PreAuthenticationGuardToken.php [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+     * @return mixed The user credentials
+     */
+    public function getCredentials()
+    {
+        return $this->credentials;
+    }
+
+    public function setAuthenticated($authenticated)
+    {
+        throw new \Exception('The PreAuthenticationGuardToken is *always* not authenticated');
```
Comment (iltar) [inline src/Symfony/Component/Security/Guard/Token/PreAuthenticationGuardToken.php]:
```
throw new \LogicException('The PreAuthenticationGuardToken is *never* authenticated.');
```

Even if you decide not to change it, the exception should have a dot in its message at the end.

**Comment by iltar** [comment]
I really love this PR! :+1: It makes things _a lot_ simpler. As you can see I have already worked my way through the changes to comment. It's really nice to see the difference between the pre and post authentication tokens.

**Comment by javiereguiluz** [comment]
:+1: congrats @weaverryan! It's a very nice move towards reducing perceived complexity of the Security component. I'd like to make two quick comments regarding the naming and code of some methods:

**1)** Are you planning to add new methods to get credentials from different sources than the `Request` object? If not, maybe the `...FromRequest` part of the method name is redundant:

```php
// redundant method name?
public function getCredentialsFromRequest(Request $request) { ... }

// acceptable alternative?
public function getCredentials(Request $request) { ... }
```

**2)** The [autoLogin() code](https://github.com/weaverryan/symfony-demo/blob/guard-auth/src/AppBundle/Controller/SecurityController.php#L45-66) looks very convoluted to me. In an ideal world, I'd like to do the following:

```php
$security->login($user);
```

I **know** that we can't do that for Symfony, but could we simplify that code as much as possible?

File: src/Symfony/Component/Security/Guard/Token/GuardTokenInterface.php (hunk 2 above)
Comment (weaverryan) [inline src/Symfony/Component/Security/Guard/Token/GuardTokenInterface.php]: Good call - I updated that and a few other bits of the words here :)

File: src/Symfony/Component/Security/Guard/Token/PreAuthenticationGuardToken.php (hunk 3 above)
Comment (weaverryan) [inline src/Symfony/Component/Security/Guard/Token/PreAuthenticationGuardToken.php]: Ah yes, thank you - I agree on both points! I've added dots, so if you see any missing now, please let me know!

File: src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+    private $rememberMeServices;
+
+    /**
+     * @param GuardAuthenticatorHandler      $guardHandler          The Guard handler
+     * @param AuthenticationManagerInterface $authenticationManager An AuthenticationManagerInterface instance
+     * @param string                         $providerKey           The provider (i.e. firewall) key
+     * @param GuardAuthenticatorInterface[]  $guardAuthenticators   The authenticators, with keys that match what's passed to GuardAuthenticationProvider
+     * @param LoggerInterface                $logger                A LoggerInterface instance
+     */
+    public function __construct(GuardAuthenticatorHandler $guardHandler, AuthenticationManagerInterface $authenticationManager, $providerKey, $guardAuthenticators, LoggerInterface $logger = null)
```
Comment (iltar) [inline src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php]: According to the docblock, `$guardAuthenticators` is an array of `GuardAuthenticatorInterface`. Should the argument be `array $guardAuthenticators`, or is this left out intentionally?

File: src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+
+    /**
+     * Iterates over each authenticator to see if each wants to authenticate the request.
+     *
+     * @param GetResponseEvent $event
+     */
+    public function handle(GetResponseEvent $event)
+    {
+        if (null !== $this->logger) {
+            $this->logger->info('Checking for guard authentication credentials', array('firewall_key' => $this->providerKey, 'authenticators' => count($this->guardAuthenticators)));
```
Comment (iltar) [inline src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php]: As related to the Exceptions, a dot is missing here. I can recall a PR where I fixed most of those logging cases where I've added dots, for consistency they could be added to all logging cases as well.

File: src/Symfony/Component/Security/Guard/Provider/GuardAuthenticationProvider.php (hunk 1 above)
Comment (weaverryan) [inline src/Symfony/Component/Security/Guard/Provider/GuardAuthenticationProvider.php]: A lot here :)

1) Actually, `checkPostAuth()` is right, but `checkPreAuth()` was wrong - it should happen right after fetching the User, but before checking credentials (see `UserAuthenticationProvider`). I've just split a method on the interface to make this possible and moved the `checkPreAuth()` in the provider. Thanks for asking about this - I hadn't really realized I had things in the wrong spot.

2) About the IP timeout idea, was this simpler with the simple-authenticator? Or are you saying that in both cases, you'd need to use inheritance or duplication to put the code into the "simple authenticator" or the "guard authenticator", so it's the same, but no better? I just commented on #11090 - I think it can be solved there.

3) We should avoid using forms or validation anywhere. But more importantly, the "guard authenticators" are so simple that you can do whatever you want, including using forms and validation.

File: src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php (hunk 5 above)
Comment (weaverryan) [inline src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php]: Ah, thanks - I hopefully have them all now!

File: src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php (hunk 4 above)
Comment (weaverryan) [inline src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php]: Not intentional at all - thanks for catching it

**Comment by weaverryan** [comment]
@javiereguiluz 

* Good idea on `checkCredentials()` - I've changed that.

* For `autoLogin`, I hear you. The system is already a big improvement over before, where you had not access to your normal "success listener Response" nor an automatic way to dispatch the "interactive login" event, which now happens. What we're missing to make it easier is "user-land" access to *which* firewall we are currently under. I'm not aware of any way to access this currently, which is why we have that ugly `secured_area` thing in there.

Thanks!

File: src/Symfony/Component/Security/Guard/Provider/GuardAuthenticationProvider.php (hunk 1 above)
Comment (iltar) [inline src/Symfony/Component/Security/Guard/Provider/GuardAuthenticationProvider.php]: Regarding point 2, It will be similar. In my case I've actually implement a custom UserChecker, which I manually had to inject (can't configure that _at all_). I've made a custom method called `checkPreUserLoad()` in a custom `UserCheckerInterface` implementation and created an abstract authenticator that all our different authentication classes (simple-form, simple-pre-auth) extend. In your implementation, I would probably put half of this inside the part where you get the values from the request and pass the captcha valid/invalid/not required from there.

I was wondering if it would be possible to tackle that problem while we are at it.

```
    public function checkPreUserLoad()
    {
        $request  = $this->request_stack->getCurrentRequest();
        $resolved = $this->login_ip_resolver->resolveStatus($request->getClientIp());

        if ($resolved->requiresBlock()) {
            throw new BlockThresholdExceededException('Too many failed login attempts, ip blocked');
        }

        if (!$resolved->requiresCaptcha()) {
            return;
        }
        // captcha is required
        $form = $this->form_factory->create('login', null, ['add_captcha' => true]);
        $form->handleRequest($request);

        // only check the captcha, there is currently no username/password validation
        if ($request->request->has('login') && $form->has('captcha')) {
            if ($form->isSubmitted() && !$form->get('captcha')->isValid()) {
                throw new InvalidCaptchaException('Provided captcha was not valid');
            }

            // captcha is valid, we can skip it
            return;
        }

        throw new CaptchaThresholdExceededException('Too many failed login attempts, captcha required');
    }
```

File: src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+    private $rememberMeServices;
+
+    /**
+     * @param GuardAuthenticatorHandler      $guardHandler          The Guard handler
+     * @param AuthenticationManagerInterface $authenticationManager An AuthenticationManagerInterface instance
+     * @param string                         $providerKey           The provider (i.e. firewall) key
+     * @param GuardAuthenticatorInterface[]  $guardAuthenticators   The authenticators, with keys that match what's passed to GuardAuthenticationProvider
+     * @param LoggerInterface                $logger                A LoggerInterface instance
+     */
+    public function __construct(GuardAuthenticatorHandler $guardHandler, AuthenticationManagerInterface $authenticationManager, $providerKey, array $guardAuthenticators, LoggerInterface $logger = null)
```
Comment (ogizanagi) [inline src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php]: Is there anything preventing you from using a `NullLogger` instance if not provided ? In order to avoid such things:
```php
if (null !== $this->logger) {
...
}
```

**Comment by ogizanagi** [comment]
Wow ! This looks great ! :+1:

File: src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php (hunk 6 above)
Comment (weaverryan) [inline src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php]: Nothing technical preventing that, but `NullLogger` isn't used anywhere in core currently - having an optional argument like this is used :)

File: src/Symfony/Component/Security/Guard/Provider/GuardAuthenticationProvider.php (hunk 1 above)
Comment (weaverryan) [inline src/Symfony/Component/Security/Guard/Provider/GuardAuthenticationProvider.php]: Would the user checker stuff in #11090 solve this? If not, what do you propose? This looks like a lot of business logic to me, and putting that either into the user checker or somewhere in the authenticator (or the authenticator calls out to another service which holds the logic) seems right and simple to me.

File: src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php (hunk 6 above)
Comment (iltar) [inline src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php]: What I think he means is the following:

```php
$this->logger = $logger ?: new NullLogger();
```
IMO this is a cleaner solution as you don't have to worry about `if($this->logger)` statements. You can just log away safely.

File: src/Symfony/Component/Security/Guard/Provider/GuardAuthenticationProvider.php (hunk 1 above)
Comment (iltar) [inline src/Symfony/Component/Security/Guard/Provider/GuardAuthenticationProvider.php]: I think the solution I proposed in #11090 will in fact solve this issue. That is Backwards compatible, just need to modify the configuration for this factory.

Eventually I can use a service decorator to replace the `security.user_checker` and voila, everything magically works without needing to change this code.

File: src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php (hunk 6 above)
Comment (ogizanagi) [inline src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php]: See #14682

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/GuardAuthenticationFactory.php [hunk 7]
```diff
@@ -0,0 +1,113 @@
+<?php
```
Comment (dunglas) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/GuardAuthenticationFactory.php]: Missing license header.

File: src/Symfony/Component/Security/Guard/Token/PreAuthenticationGuardToken.php [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+     * @return mixed The user credentials
+     */
+    public function getCredentials()
+    {
+        return $this->credentials;
+    }
+
+    public function setAuthenticated($authenticated)
+    {
+        throw new \LogicException('The PreAuthenticationGuardToken is *always* not authenticated.');
```
Comment (WouterJ) [inline src/Symfony/Component/Security/Guard/Token/PreAuthenticationGuardToken.php]: I know it's quite common in English, but being dutch, "always not" is very confusing for me. What about replacing it with "never"?

Comment (nesl247) [inline src/Symfony/Component/Security/Guard/Token/PreAuthenticationGuardToken.php]: Either `is *never* authenticated`, or `is *always* unauthenticated`.

File: src/Symfony/Component/Security/Guard/Provider/GuardAuthenticationProvider.php [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+    private $providerKey;
+    private $userChecker;
+
+    /**
+     * @param GuardAuthenticatorInterface[] $guardAuthenticators The authenticators, with keys that match what's passed to GuardAuthenticationListener
+     * @param UserProviderInterface         $userProvider        The user provider
+     * @param string                        $providerKey         The provider (i.e. firewall) key
+     * @param UserCheckerInterface          $userChecker
+     */
+    public function __construct(array $guardAuthenticators, UserProviderInterface $userProvider, $providerKey, UserCheckerInterface $userChecker)
```
Comment (WouterJ) [inline src/Symfony/Component/Security/Guard/Provider/GuardAuthenticationProvider.php]: 99 out of 100 times, the user checker passed through here is the default `UserChecker`, right? What do you think about making this final argument optional and defaulting to the default user checker? E.g. `$this->userChecker = $userChecker ?: new UserChecker();`

File: src/Symfony/Component/Security/Guard/Provider/GuardAuthenticationProvider.php [hunk 10]
```diff
@@ -0,0 +1,140 @@
+<?php
+
+namespace Symfony\Component\Security\Guard\Provider;
```
Comment (WouterJ) [inline src/Symfony/Component/Security/Guard/Provider/GuardAuthenticationProvider.php]: I propose to move this namespace into `Guard\Authentication`, just as done in Security Core. Or move the class in `Guard\Authentication` directly (to avoid long FQCNs). The -AuthenticationProvider prefix already makes it clear it's an authentication provider, no need to duplicate that in the namespaces imo (in my ideal world, it would be `Guard\Authentication\Provider\Guard`).

File: src/Symfony/Component/Security/Guard/Provider/GuardAuthenticationProvider.php [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+                get_class($guardAuthenticator),
+                is_object($user) ? get_class($user) : gettype($user)
+            ));
+        }
+
+        // check the preAuth UserChecker
+        $this->userChecker->checkPreAuth($user);
+        // check the credentials
+        $guardAuthenticator->checkCredentials($token->getCredentials(), $user);
+        // check the postAuth UserChecker
```
Comment (WouterJ) [inline src/Symfony/Component/Security/Guard/Provider/GuardAuthenticationProvider.php]: I don't think these comments explain anything more (as it's just `humanize(method_name) property_name`).

File: src/Symfony/Component/Security/Guard/Provider/GuardAuthenticationProvider.php [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+        if (null === $user) {
+            throw new UsernameNotFoundException(sprintf(
+                'Null returned from %s::getUser()',
+                get_class($guardAuthenticator)
+            ));
+        }
+
+        if (!$user instanceof UserInterface) {
+            throw new \UnexpectedValueException(sprintf(
+                'The %s::getUser method must return a UserInterface. You returned %s.',
```
Comment (WouterJ) [inline src/Symfony/Component/Security/Guard/Provider/GuardAuthenticationProvider.php]: to be consistent, use `::getUser()`

File: src/Symfony/Component/Security/Guard/Provider/GuardAuthenticationProvider.php [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+        // check the credentials
+        $guardAuthenticator->checkCredentials($token->getCredentials(), $user);
+        // check the postAuth UserChecker
+        $this->userChecker->checkPostAuth($user);
+
+        // turn the UserInterface into a TokenInterface
+        $authenticatedToken = $guardAuthenticator->createAuthenticatedToken($user, $this->providerKey);
+        if (!$authenticatedToken instanceof TokenInterface) {
+            throw new \UnexpectedValueException(sprintf(
+                'The %s::createAuthenticatedToken method must return a TokenInterface. You returned %s.',
```
Comment (WouterJ) [inline src/Symfony/Component/Security/Guard/Provider/GuardAuthenticationProvider.php]: `::createAuthenticatedToken()`

File: src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+     *
+     * @param GuardAuthenticatorInterface $guardAuthenticator
+     * @param Request                     $request
+     * @param TokenInterface              $token
+     * @param Response                    $response
+     */
+    private function triggerRememberMe(GuardAuthenticatorInterface $guardAuthenticator, Request $request, TokenInterface $token, Response $response = null)
+    {
+        if (!$guardAuthenticator->supportsRememberMe()) {
+            return;
```
Comment (sstok) [inline src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php]: Should you not inform that the guard doesn't support rembember-me? And do this check *after* checking if remember-me was actually enabled for the firewall.

File: src/Symfony/Component/Security/Guard/AbstractGuardAuthenticator.php [hunk 15]
```diff
@@ -0,0 +1,32 @@
+<?php
+
```
Comment (dupuchba) [inline src/Symfony/Component/Security/Guard/AbstractGuardAuthenticator.php]: Licence header are missing here too :-)

File: src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php (hunk 6 above)
Comment (dupuchba) [inline src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php]: agree with @iltar it does simplify a little bit IMO

File: src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php [hunk 16]
```diff
@@ -0,0 +1,180 @@
...
+        if (null !== $this->logger) {
+            $this->logger->info('Checking for guard authentication credentials.', array('firewall_key' => $this->providerKey, 'authenticators' => count($this->guardAuthenticators)));
+        }
+
+        foreach ($this->guardAuthenticators as $key => $guardAuthenticator) {
+            // get a key that's unique to *this* guard authenticator
+            // this MUST be the same as GuardAuthenticationProvider
+            $uniqueGuardKey = $this->providerKey.'_'.$key;
+
+            $this->executeGuardAuthenticator($uniqueGuardKey, $guardAuthenticator, $event);
```
Comment (Taluu) [inline src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php]: instead of injecting a `uniqueGuardKey`, how about having a `getKey` method on the guard authenticator, and use that key with the `providerKey` in the `executeGuardAuthenticator` body ? Granted, this is preventing from having the same authenticator registered several times, and this should be good too, nope ?

**Comment by OskarStark** [comment]
great work :+1:

**Comment by ihsanudin** [comment]
very nice :+1:

**Comment by malas** [comment]
I waited for this so long! Hope it gets into core ASAP

**Comment by giovkanata** [comment]
Well done! This was one of the most lacked features in symfony until now.

**Comment by nielvrom** [comment]
Very nice! In the beginning of learning Symfony the authentication was not easy for me! This makes it so much easier for new users!

**Comment by pbowyer** [comment]
Awesome work @weaverryan and something Symfony2 has long needed. Human Focused Development!

File: src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php (hunk 6 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php]: :-1: for a NullLogger, Symfony should take options for most performance.

File: src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php (hunk 16 above)
Comment (weaverryan) [inline src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php]: Yes, it would remove this hack-ish looking code, but it's a meaningless key that the user shouldn't need to bother with. And you could have a single authenticator that is able to be configured (via constructor params) in different ways, so that it *does* make sense to have 2 at once. Anyways, we'll see if others agree with you.

File: src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php (hunk 14 above)
Comment (weaverryan) [inline src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php]: Sounds reasonable to me :) making that change

File: src/Symfony/Component/Security/Guard/Provider/GuardAuthenticationProvider.php (hunk 10 above)
Comment (weaverryan) [inline src/Symfony/Component/Security/Guard/Provider/GuardAuthenticationProvider.php]: Finally updating this PR :). I don't understand what you're proposing here - when you suggest `Guard\Authentication...` - do you actually mean that, or actually `Symfony\Component\Security\Guard\Authentication...`? I'm not convinced I have the classes organized in the most efficient way, and it's an easy change to still make.

File: src/Symfony/Component/Security/Guard/Provider/GuardAuthenticationProvider.php (hunk 9 above)
Comment (weaverryan) [inline src/Symfony/Component/Security/Guard/Provider/GuardAuthenticationProvider.php]: I don't know if I like adding a `new UserChecker` automatically, but I think I would be for making the argument optional (and just checking for the UserChecker before calling it below). I know your thought is for making these classes somewhat usable directly.

**Comment by weaverryan** [comment]
Ping @symfony/deciders

This is a large feature and security is complex. For that reason, I released the KnpUGuard library 3+ months ago to get proven and work out bugs. It's gotten 1300+ installs, and I've talked with people who have used it to integrate LDAP, OAuth, API tokens auth, multiple firewalls, etc. So, fwiw, the tires have been kicked.

Thanks!

**Comment by fabpot** [comment]
Not yet reviewed the code, but I have one suggestion: the guard sub-system is well contained into its own sub-namespace. I can see how this new system could be used by Silex as well, but to be able to do that I would need to depend on `symfony/security`. As all other sub-parts of the security component are available as independent Composer packages, I think it would make sense to do the same here. @weaverryan Can you make the needed changes?

**Comment by fabpot** [comment]
Regarding the `simple_*` classes added a while back, I think we need to deprecate them; having 3 different ways to do the same thing is not a good idea. So, having a high level one like this one and a low-level one is probably more than enough.

**Comment by fabpot** [comment]
:+1:

**Comment by iltar** [comment]
I agree with @fabpot, but would that be needed for 2.8/3.0 already? In my opinion this will make a too big of a BC. In my case, this would add an additional month (minimal) of time to upgrade because it will need a complete rewrite.

Maybe this is something that can be deprecated in 3.1 and removed in 4.0? The security component is quite complex so it would be nice to give people a lot of time to figure out this part.

**Comment by fabpot** [comment]
@iltar fair enough; just created an issue to not forget to do that in 3.1 (see #15873)

**Comment by weaverryan** [comment]
I like the deprecation plan. We can replace simple* in the documentation with Guard for 2.8.

I've added the pieces to make Guard its own component - but there's a good chance I foo-barred something with the versions in composer.json. The dependencies on the other security components could probably have lower versions, but do we care?

**Comment by fabpot** [comment]
Thank you @weaverryan.

**Comment by fabpot** [comment]
The subtree-split is now active for this repository and available here:

https://github.com/symfony/security-guard
https://packagist.org/packages/symfony/security-guard

**Comment by derrabus** [comment]
@weaverryan Symfony 2.7 compatibility in composer.json wouldn't be too bad, so I could try out the guard in an existing application without pulling the entire framework from an unstable branch.

Apart from that: :clap: Great work, I'm glad to see this feature being merged. :smiley:

**Comment by derrabus** [comment]
The unit tests pass, if I pull every required Symfony component from 2.7.4 but security-core. As far as I can tell, the only reason why security-core has to be `~2.8@dev` is the usage of a new exception class (`AuthenticationExpiredException`). I'm not sure if you would want to work around this.

But what you can probably do is setting security-http to `~2.7`.

**Comment by TomasVotruba** [comment]
@weaverryan Great job!

**Comment by ihsanudin** [comment]
thanks @fabpot :+1:

File: src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php (hunk 16 above)
Comment (lyrixx) [inline src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php]: (It's a bit late I know, But I lacked time to review it)

IIUC the code, the last guardAuthenticator have an higher precedence than other one. 
I say that because all guardAuthenticators are execute one after an other, and all are able to 
set a new response in the event and because you never check if a response it set.

That mean if the first guardAuthenticator allow an user, the second one could reject the user.
So finally, the user will be rejected. I'm not sure that is what end developer wants.

File: src/Symfony/Component/Security/Guard/Tests/Firewall/GuardAuthenticationListenerTest.php [hunk 17]
```diff
@@ -0,0 +1,222 @@
...
+    }
+
+    protected function tearDown()
+    {
+        $this->authenticationManager = null;
+        $this->guardAuthenticatorHandler = null;
+        $this->event = null;
+        $this->logger = null;
+        $this->request = null;
+    }
```
Comment (lyrixx) [inline src/Symfony/Component/Security/Guard/Tests/Firewall/GuardAuthenticationListenerTest.php]: Miss `$this->rememberMeServices = null`

Comment (weaverryan) [inline src/Symfony/Component/Security/Guard/Tests/Firewall/GuardAuthenticationListenerTest.php]: addressed in #15920

**Comment by weaverryan** [comment]
@derrabus I've opened #15920 to lower the http requirement. But I don't think we'll want to add a `class_exists` with the `AuthenticationExpiredException` to get 2.7 compatibility

File: src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php (hunk 16 above)
Comment (weaverryan) [inline src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php]: Ah, very interesting - and yes, you do understand the code well.

With normal security, if one listener fails, then the other ones aren't called, because `$event->setResponse()` is called with the entry point.  But on success, it depends: some listeners like `UsernamePasswordFormAuthenticationListener` set a response on the event, so nobody else would be called. But others - like `BasicAuthenticationListener` don't set a response. So on success, another authentication listener could be called, which could fail and reject the user (just like you're saying). In other words, this is already the behavior of some core authentication listeners.

See #15925 - I've updated to return on the Event having a response, which should make it fully mimic core. At least with guard, you can control the "priority" of the authenticators, in the edge case that you care.

**Comment by Rvanlaak** [comment]
Nice! Would this also allow "simple" configuration via annotations? https://github.com/symfony/symfony/issues/13950

**Comment by Codenator81** [comment]
@weaverryan Super simple! Love it!

File: src/Symfony/Component/Security/Guard/GuardAuthenticatorInterface.php [hunk 18]
```diff
@@ -0,0 +1,150 @@
...
+     * Throw an AuthenticationException if the credentials are invalid.
+     *
+     * The *credentials* are the return value from getCredentials()
+     *
+     * @param mixed         $credentials
+     * @param UserInterface $user
+     *
+     * @throws AuthenticationException
+     */
+    public function checkCredentials($credentials, UserInterface $user);
```
Comment (patrick-mcdougle) [inline src/Symfony/Component/Security/Guard/GuardAuthenticatorInterface.php]: It makes be nervous that this method, **by default, results in a user being successfully authenticated** and requires throwing an exception in the case that the user should *NOT* be authenticated. In the case that a developer accidentally writes a bug that doesn't throw an exception in a complex (or any) setup, the mistake results in ***users being authenticated that shouldn't have been.***

I think a safer way to do this would be to explicitly return a value (perhaps a boolean) indicating whether the credential check was successful or not. In the `GuardAuthenticationProvider` you could very easily check this boolean value for falseness (in the case that someone doesn't return) and then throw an `AuthenticationException` to keep the rest of the logic in the Provider the same.

TLDR; Which is worse? **Authenticating a user who has invalid credentials?** Or not authenticating a user who has valid credentials? I guess I'd rather not take a chance with security. Thoughts?

Comment (dunglas) [inline src/Symfony/Component/Security/Guard/GuardAuthenticatorInterface.php]: :+1:

**Comment by iltar** [comment]
- If you return a boolean, how will you know what went wrong? 
 - How will you be able to trace back the reason of not being able to login, to this specific method implementation?

Imo this is a developer mistake and tests should've caught that. There's already a weird case with `ParamConverters` having to return a boolean for some reason which makes it far more confusion than necessary.

As long as it's stated that an exception should be thrown in the docs: :-1: from my side for this request.

Maybe result objects are a better suggestion, where you can specify a reason. It would conflict with the current flow of `AuthenticationException` thought.

File: src/Symfony/Component/Security/Guard/GuardAuthenticatorInterface.php (hunk 18 above)
Comment (webmozart) [inline src/Symfony/Component/Security/Guard/GuardAuthenticatorInterface.php]: Not sure I agree - the same could be said for returning `false`. For example, if you collect the result in a boolean variable `$valid` which is returned, that variable could easily have the wrong truth value simply by mistaking `&&` for `||` somewhere in your code.

The only way to prevent bugs like the one you described is by properly testing the authenticator. I don't think that throwing an exception vs. returning `false` makes any difference here.

Comment (fabpot) [inline src/Symfony/Component/Security/Guard/GuardAuthenticatorInterface.php]: see #16395

Comment (patrick-mcdougle) [inline src/Symfony/Component/Security/Guard/GuardAuthenticatorInterface.php]: Heh, maybe I should look at the tip of 2.8 before I go commenting on old PRs. Sorry everyone.

Comment (weaverryan) [inline src/Symfony/Component/Security/Guard/GuardAuthenticatorInterface.php]: @patrick-mcdougle No, thank you! I was trying to find you at the conference - I made a pull request within an hour after your asked this very good question - and wanted to tell you about it :).

**Comment by bogdaniel** [comment]
Question in your working example shouldn't be return true if else?
File: [FormLoginAuthenticator.php](https://github.com/weaverryan/symfony-demo/blob/guard-auth/src/AppBundle/Security/FormLoginAuthenticator.php)
[Line: 60](https://github.com/weaverryan/symfony-demo/blob/guard-auth/src/AppBundle/Security/FormLoginAuthenticator.php#L60)

```
        if (!$passwordValid) {
            throw new BadCredentialsException();
        }
```
To
```
        if ($passwordValid)
            return true;
        else
            throw new BadCredentialsException();
```

Because i think it will always hit the user with a BadCredentialsException()

**Comment by WouterJ** [comment]
@bogdaniel this behaviour was changed after this PR (but before releasing 2.8.0) it seems like the demo hasn't been updated.

**Comment by derrabus** [comment]
@WouterJ This shouldn't be the place to maintain an up-to-date demo for this feature. imho, the PR should be seen as a documentation of the development process and the demo shows how the feature has worked at the time when the PR was merged.

@bogdaniel: Please see the docs for an up-to-date documentation on this feature: http://symfony.com/doc/current/cookbook/security/guard-authentication.html
